File: src/main/java/clustering/SingleLinkHAC.java
package clustering;

import core.DocId;
import core.zones.Zone;
import core.Indexer;

import java.util.*;

import static core.vectorizer.VectorUtils.getCosineAngle;

public class SingleLinkHAC {
    public static SingleLinkHAC fromIndexer(Indexer indexer) {
        return new SingleLinkHAC(indexer);
    }

    private final Indexer indexer;

    private SingleLinkHAC(Indexer indexer) {
        this.indexer = indexer;
    }

    public Set<DocumentCluster> buildCluster(double minSimilarity) {
        return buildCluster(minSimilarity, -1);
    }

    public Set<DocumentCluster> buildCluster(double minSimilarity, int iterations) {
        final int numDocuments = indexer.getDocIds().size();

        PriorityQueue<PairSimilarity> similarityQueue = new PriorityQueue<>(numDocuments * numDocuments, new Comparator<PairSimilarity>() {
            @Override
            public int compare(PairSimilarity o1, PairSimilarity o2) {
                return Double.compare(o2.score, o1.score);
            }
        });

        // Initialize similarity table.
        double[][] similarityTable = new double[numDocuments][numDocuments];
        for (int i = 0; i < similarityTable.length; i++) {
            for (int j = 0; j < similarityTable.length; j++) {
                similarityTable[i][j] = 0.0;
            }
        }

        // Build initial cluster map.
        Map<Integer, DocumentCluster> documentClusterMap = new HashMap<>();
        for (int i = 0; i < numDocuments; i++) {
            DocId docId = indexer.getDocIds().get(i);

            DocumentCluster documentCluster = new DocumentCluster();
            documentCluster.add(docId);

            documentClusterMap.put(i, documentCluster);
        }

        // Build similarity table.
        for (int i = 0; i < numDocuments; i++) {
            System.out.printf("Building similarity table: %d/%d\n", i, numDocuments);
            for (int j = i; j < numDocuments; j++) {
                for (Zone zone : indexer.getZones()) {
                    double similarity = (i == j) ? 1.0 : getCosineAngle(
                            zone.getVector(indexer.getDocIds().get(i)),
                            zone.getVector(indexer.getDocIds().get(j)));

                    if (similarity < minSimilarity) similarity = 0;

                    similarityTable[i][j] = similarity;
                    similarityTable[j][i] = similarity;

                    similarityQueue.add(new PairSimilarity(i, j, similarity));

                    if (i == 0) {
                        System.out.printf("Vectorizing: %d/%d\n", j, numDocuments);
                    }
                }
            }
        }

        int iteration = 0;
        Set<Integer> mergedClusters = new HashSet<>();
        boolean isClustering = true;
        while (isClustering) {
            iteration++;

            Set<Integer> activeClusters = new HashSet<>();
            List<PairSimilarity> activeMerges = new ArrayList<>();

            // Find merge-able clusters.
            double maxSimilarity = -1;

            while (!similarityQueue.isEmpty()) {
                PairSimilarity pairSimilarity = similarityQueue.poll();

                if (activeClusters.contains(pairSimilarity.i) || activeClusters.contains(pairSimilarity.j)) continue;
                if (mergedClusters.contains(pairSimilarity.i) || mergedClusters.contains(pairSimilarity.j)) continue;

                if (pairSimilarity.i == pairSimilarity.j) continue;

                if (pairSimilarity.score > minSimilarity) {
                    activeClusters.add(pairSimilarity.i);
                    activeClusters.add(pairSimilarity.j);

                    activeMerges.add(pairSimilarity);

                    if (maxSimilarity < 0) maxSimilarity = pairSimilarity.score;
                }
            }

            // Update similarity table.
            for (PairSimilarity mergePair : activeMerges) {
                int i = mergePair.i;
                int j = mergePair.j;

                for (int ndx = 0; ndx < numDocuments; ndx++) {
                    double updatedSimilarity = Math.max(similarityTable[i][ndx], similarityTable[j][ndx]);
                    similarityTable[i][ndx] = updatedSimilarity;
                    similarityTable[ndx][i] = updatedSimilarity;
                    similarityTable[j][ndx] = updatedSimilarity;
                    similarityTable[ndx][j] = updatedSimilarity;

                    similarityQueue.add(new PairSimilarity(j, ndx, updatedSimilarity));
                }
            }

            // Merge active clusters.
            for (PairSimilarity mergePair : activeMerges) {
                int sourceNdx = mergePair.i;
                int targetNdx = mergePair.j;

                DocumentCluster sourceCluster = documentClusterMap.get(sourceNdx);
                DocumentCluster targetCluster = documentClusterMap.get(targetNdx);

                targetCluster.merge(sourceCluster);

                documentClusterMap.remove(sourceNdx);

                mergedClusters.add(sourceNdx);
            }

            isClustering = iterations != -1 && iteration < iterations && !activeClusters.isEmpty();

            System.out.printf("Iter %d: Max score: %f\n", iteration, maxSimilarity);
        }

        return new HashSet<>(documentClusterMap.values());
    }

    private static class PairSimilarity {
        public int i;
        public int j;
        public double score;

        private PairSimilarity(int i, int j, double score) {
            this.i = i;
            this.j = j;
            this.score = score;
        }
    }
}

File: src/main/java/clustering/singlelinkhac/SingleLinkHAC.kt
package clustering.singlelinkhac

import clustering.DocumentCluster
import core.DocId
import core.Record
import core.zones.Zone
import java.util.HashSet
import kotlin.math.max

class SingleLinkHAC private constructor(private val zones: HashSet<Zone<String>>) {
    fun buildCluster(records: List<Record>, minSimilarity: Float): Set<DocumentCluster> {
        // Index.
        val zonesByName = zones.associateBy { it.name }
        for (record in records) {
            record.dataByZoneName.entries.map { zoneData ->
                val zoneToIndex = zonesByName[zoneData.key] ?: throw RuntimeException("Invalid zone index")
                zoneToIndex.index(zoneData.value as String, record.docId)
            }
        }

        // Build similarity table.
        val similarityArray = Array<Array<Double>>(records.size) { Array(records.size, { 0.toDouble() }) }
        for (i in records.indices) {
            System.out.printf("Building similarity: [%d/%d]\n", i, records.size)

            val record = records[i]
            val similarities = record.dataByZoneName.entries
                    .flatMap { zoneData -> zonesByName[zoneData.key]?.matchData(zoneData.value as String) ?: emptyList() }
                    .groupBy ({ it.first }, { it.second })
                    .mapValues { it.value.fold(1.0) { acc, score -> acc * score } }

            for (j in records.indices) {
                similarityArray[i][j] = similarities[records[j].docId] ?: 0.toDouble()
                similarityArray[j][i] = similarities[records[j].docId] ?: 0.toDouble()
            }
        }

        // Build clusters.
        val clusters = mutableMapOf<Int, DocumentCluster>()
        for (i in records.indices) {
            val record = records[i]
            clusters.put(i, DocumentCluster().apply { add(record.docId) })
        }

        var iteration = 0
        var maxSimilarity = 0.toDouble()

        var allowClustering = true
        while (allowClustering) {
            iteration = iteration + 1

            // Find merge-able clusters.
            val activeClusters = mutableSetOf<Int>()
            val merges = mutableListOf<Pair<Int, Int>>()
            for (i in similarityArray.indices) {
                val maxPair = findMaxSimilarity(similarityArray, activeClusters, clusters)

                // No eligible pairs found.
                if (maxPair.first < 0 || maxPair.second < 0) break

                // Minimum threshold not met.
                maxSimilarity = max(similarityArray[maxPair.first][maxPair.second], maxSimilarity)
                if (similarityArray[maxPair.first][maxPair.second] < minSimilarity) break

                merges.add(maxPair)
                activeClusters.add(maxPair.first)
                activeClusters.add(maxPair.second)
            }

            // Update similarity table.
            for (merge in merges) {
                for (i in similarityArray.indices) {
                    similarityArray[merge.first][i] = max(similarityArray[merge.first][i], similarityArray[merge.second][i])
                    similarityArray[i][merge.first] = similarityArray[merge.first][i]

                    similarityArray[merge.second][i] = similarityArray[merge.first][i]
                    similarityArray[i][merge.first] = similarityArray[merge.first][i]
                }
            }

            // Merge clusters.
            for (merge in merges) {
                val clusterIn = clusters[merge.second]
                        ?: throw RuntimeException("Cluster not found! Should not happen.")
                val clusterOut = clusters[merge.first]
                        ?: throw RuntimeException("Cluster not found! Should not happen.")

                clusterOut.merge(clusterIn)

                clusters[merge.second] = clusterOut
            }

            allowClustering = activeClusters.isNotEmpty()

            System.out.printf("Iter %d: Max score: %f\n", iteration, maxSimilarity)
        }

        return clusters.values.toSet()
    }

    private fun findMaxSimilarity(similarities: Array<Array<Double>>, ignoreIndices: Set<Int>, clusters: Map<Int, DocumentCluster>): Pair<Int, Int> {
        var maxValue: Double = 0.toDouble()
        var maxPair = Pair(-1, -1)
        for (i in similarities.indices) {
            for (j in similarities.indices) {
                if (i == j) continue

                if (ignoreIndices.contains(i) || ignoreIndices.contains(j)) continue

                if (clusters[i]?.equals(clusters[j]) ?: true) continue

                if (similarities[i][j] > maxValue) {
                    maxValue = similarities[i][j]
                    maxPair = Pair(i, j)
                }
            }
        }

        return maxPair
    }

    private data class DocumentSimilarity constructor(val docIdA: DocId, val docIdB: DocId, val score: Double)

    class Builder {
        private val zones = HashSet<Zone<String>>()

        fun addZone(zone: Zone<String>): Builder {
            this.zones.add(zone)
            return this
        }

        fun build(): SingleLinkHAC {
            return SingleLinkHAC(zones)
        }
    }
}
